sgkit/stats/pedigree.py

```diff
@@ -649,3 +649,368 @@ def pedigree_kinship(
+    ds = define_variable_if_absent(
+        ds,
+        variables.stat_pedigree_kinship,
+        stat_pedigree_kinship,
+        pedigree_kinship,
+        method=method,
+        stat_Hamilton_Kerr_tau=stat_Hamilton_Kerr_tau,
+    )
+    variables.validate(
+        ds, {stat_pedigree_kinship: variables.stat_pedigree_kinship_spec}
+    )
+    kinship = ds[stat_pedigree_kinship].data
+    if method == "diploid":
+        A = kinship * 2
+    elif method == "Hamilton-Kerr":
+        ploidy = ds[stat_Hamilton_Kerr_tau].data.sum(axis=-1)
+        A = da.apply_gufunc(
+            kinship_as_additive_relationship,
+            "(),(),()->()",
+            kinship,
+            ploidy[None, :],
+            ploidy[:, None],
+        )
+    new_ds = create_dataset(
+        {
+            variables.stat_additive_relationship: xr.DataArray(
+                A, dims=["samples_0", "samples_1"]
+            ),
+        }
+    )
+    return conditional_merge_datasets(ds, new_ds, merge)
+
+
+@njit(cache=True)
+def _update_inverse_additive_relationships(
+    mtx: ArrayLike,
+    kinship: ArrayLike,
+    parent: ArrayLike,
+    i: int,
+    tau: Union[ArrayLike, None] = None,
+) -> None:  # pragma: no cover
+    p, q = parent[i]
+    if tau is None:
+        # assume diploid
+        tau_p, tau_q = 1, 1
+    else:
+        tau_p, tau_q = tau[i, 0], tau[i, 1]
+    # weighted contribution of each parent
+    ploidy_i = tau_p + tau_q
+    weight_p, weight_q = tau_p / ploidy_i, tau_q / ploidy_i
+    # sparse matrix product of weight vectors and kinship matrix
+    prod = 0.0
+    if p >= 0:
+        prod += kinship[p, p] * weight_p**2
+    if q >= 0:
+        prod += kinship[q, q] * weight_q**2
+    if (p >= 0) and (q >= 0):
+        prod += kinship[p, q] * weight_p * weight_q * 2
+    scalar = 1 / (kinship[i, i] - prod)
+    # Calculate inverse kinships and adjust to inverse relationships.
+    # using sparse matrix multiplication.
+    # The inverse kinships are then adjusted to inverse relationships by
+    # dividing by a weighting of the ploidy of each pair of individuals.
+    if p >= 0:
+        ploidy_p = 2 if tau is None else tau[p, 0] + tau[p, 1]
+        mtx[p, p] += (weight_p**2 * scalar) / ploidy_p
+        num = weight_p * scalar
+        denom = 2 * np.sqrt(ploidy_p / 2 * ploidy_i / 2)
+        frac = num / denom
+        mtx[p, i] -= frac
+        mtx[i, p] -= frac
+    if q >= 0:
+        ploidy_q = 2 if tau is None else tau[q, 0] + tau[q, 1]
+        mtx[q, q] += (weight_q**2 * scalar) / ploidy_q
+        num = weight_q * scalar
+        denom = 2 * np.sqrt(ploidy_q / 2 * ploidy_i / 2)
+        frac = num / denom
+        mtx[q, i] -= frac
+        mtx[i, q] -= frac
+    if (p >= 0) and (q >= 0):
+        num = weight_p * weight_q * scalar
+        denom = 2 * np.sqrt(ploidy_p / 2 * ploidy_q / 2)
+        frac = num / denom
+        mtx[p, q] += frac
+        mtx[q, p] += frac
+    mtx[i, i] += scalar / ploidy_i
+    return
```

Comment:
I find the empty return odd Pythonically, but perhaps there's a good reason from a jitting perspective?